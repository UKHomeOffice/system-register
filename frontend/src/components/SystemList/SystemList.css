.system-list-filter {
  display: inline-flex;
  margin-top: 20px;
}

.system-list-dropdown {
    margin-left: 40px;
    margin-top: 16px;
    min-width: 500px;
  }

  .system-list-search{
    margin-top: 32px;
  }