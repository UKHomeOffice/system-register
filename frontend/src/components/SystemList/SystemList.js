/* eslint-disable */
import './SystemList.css'
import React, { useState, useEffect } from 'react'
import TextField from '@material-ui/core/TextField'
import MenuItem from "@material-ui/core/MenuItem";
import Select from "@material-ui/core/Select";
import SystemCard from "../SystemCard/SystemCard"

//todo make stateless, all components stateless all, state should be in the containers only
const SystemList = props => {
    const systems = props.register.systems
    const [visibleSystems, setVisibleSystems] = useState(systems)
    const [selectedPortfolio, setSelectedPortfolio] = useState('all')
    const [searchString, setSearchString] = useState('')
    const portfolios = [...new Set(systems?.map(s => s.portfolio))].sort()

    const filter = () => {
        setVisibleSystems(
            systems?.filter((s) => matchesPortfolio(s, selectedPortfolio))
                    .filter((s) => matchesSearch(s, searchString))
        )
    }

    useEffect(filter, [searchString, selectedPortfolio, props.register])

    return (
        <div className="centerContent">
            <div className="topMarginBig">
                <strong>Number of systems identified: </strong>{systems?.length}
            </div>
            <div className="bottomMargin topMarginBig">
                <TextField
                    id="search"
                    label="search"
                    onChange={e => setSearchString(e.target.value)}
                />
                <Select
                    value={selectedPortfolio}
                    onChange={e => setSelectedPortfolio(e.target.value)}
                    className="dropdown">
                    <MenuItem value="all">
                        All Portfolios
                    </MenuItem>
                    {portfolios.map((p, i) => <MenuItem key={i} value={p}>{p}</MenuItem>)}
                </Select>
            </div>
            {/* todo refactor out to functional <Systems /> component */}
            <div className="topMarginBig">
                {
                    visibleSystems?.map((system, key) => <SystemCard key={key} system={system} />)
                    /* todo use id rather than key to help out the shadow DOM */
                }
            </div>
        </div >
    )
}

function containsCaseInsensitive(a, b) {
    return a.toLowerCase().includes(b.toLowerCase())
}

function matchesPortfolio(system, portfolio) {
    return portfolio === "all" || system.portfolio === portfolio
}

function matchesSearch(system, needle) {
    const result = containsCaseInsensitive(system.name, needle) ||
        system.aliases.some((alias) => containsCaseInsensitive(alias, needle))
    return result
}

export default SystemList
