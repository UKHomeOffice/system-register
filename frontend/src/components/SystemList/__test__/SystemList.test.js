import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import SystemList from '../SystemList';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

const test_data = {
    systems: [
        {
            name: "System 1",
            portfolio: 'Portfolio 1',
            aliases: ['Alias A'],
            last_updated: {}
        },
        {
            name: "System 3",
            portfolio: 'Portfolio 2',
            aliases: ['Alias B', 'Alias C'],
            last_updated: {}
        },
        {
            name: "System 2",
            portfolio: 'Portfolio 2',
            aliases: [],
            last_updated: {}
        }
    ]
}

let _getByLabelText, _queryByText, _getByTestId = undefined

//tests
beforeEach(setup)
afterEach(cleanup)

describe('<SystemList />', () => {
    describe('searching by name', () => {
        it('renders search, filter and all systems when loaded', () => {
            expect(_getByLabelText('search')).toBeInTheDocument()
            expect(_getByTestId("select-portfolio")).toBeInTheDocument()
            expectAllSystems()
        })

        it('finds no systems when nothing matches the search string', () => {
            whenUserSearchesFor("System 99")
            expectNoSystems()
        })

        it('shows all systems when search string removed', () => {
            whenUserSearchesFor("System 1")
            expectSystems("System 1")
            whenUserSearchesFor("")
            expectAllSystems()
        })

        it('finds a system which fully match search string', () => {
            whenUserSearchesFor("System 1")
            expectSystems("System 1")
        })

        it('finds a system which partially matches search string', () => {
            whenUserSearchesFor("ystem 1")
            expectSystems("System 1")
        })

        it('finds multiple systems which match search string', () => {
            whenUserSearchesFor("System")
            expectSystems("System 1", "System 2", "System 3")
        })

        it('ignores case when searching', () => {
            whenUserSearchesFor("system 2")
            expectSystems("System 2")
        })
    })

    describe('searching by alias', () => {
        it('finds a system whos alias fully matches the search string', () => {
            whenUserSearchesFor("Alias A")
            expectSystems("System 1")
        })

        it('finds all system whos alias partialy matches the search string', () => {
            whenUserSearchesFor("Alias")
            expectSystems("System 1", "System 3")
        })

        it('finds no system when no aliases match the search string', () => {
            whenUserSearchesFor("Alias Z")
            expectNoSystems()
        })

        it('finds a system whos alias in lowercase matches the search string', () => {
            whenUserSearchesFor("alias a")
            expectSystems("System 1")
        })
    })

    describe('filtering by Portfolio', () => {
        it('filters systems when one system matches', () => {
            whenUserSelectsPortfolio("Portfolio 1")
            expectSystems("System 1")
        })

        it('filters systems when multiple systems match', () => {
            whenUserSelectsPortfolio("Portfolio 2")
            expectSystems("System 2", "System 3")
        })
    })

    describe('portfolio filter and search terms combined', () => {
        it('correctly conbines filters to return a system', () => {
            whenUserSelectsPortfolio("Portfolio 2")
            whenUserSearchesFor("3")
            expectSystems("System 3")
        })

        it('correctly conbines filters to return no systems', () => {
            whenUserSelectsPortfolio("Portfolio 2")
            whenUserSearchesFor("1")
            expectNoSystems()
        })
    })

    it('matches snapshot', () => {
        expect(renderer
            .create(<BrowserRouter><SystemList register={test_data} /></BrowserRouter>))
            .toMatchSnapshot()
    })
})

//dsl
function setup() {
    const { getByLabelText, queryByText, getByTestId } =
        render(<BrowserRouter><SystemList register={test_data} /></BrowserRouter>)
    _getByLabelText = getByLabelText
    _queryByText = queryByText
    _getByTestId = getByTestId
}

function expectSystems(...systemNames) {
    const visible = test_data.systems.filter(s => systemNames.includes(s.name))
    const invisible = test_data.systems.filter(s => !systemNames.includes(s.name))
    visible.forEach(s => expect(_queryByText(s.name)).toBeInTheDocument())
    invisible.forEach(s => expect(_queryByText(s.name)).not.toBeInTheDocument())
}

function expectNoSystems() { expectSystems() }
function expectAllSystems() { expectSystems(...test_data.systems.map(s => s.name)) }

function whenUserSearchesFor(str) {
    fireEvent.change(_getByLabelText('search'), { target: { value: str } })
}
function whenUserSelectsPortfolio(portfolio) {
    fireEvent.change(_getByTestId("select-portfolio"), { target: { value: portfolio } })
}

//mocks
jest.mock("@govuk-react/select", () => ({ value, onChange }) => {
    const options = [{ value: "Portfolio 1" }, { value: "Portfolio 2" }, { value: "Portfolio 3" }]
    function handleChange(event) {
        const option = options.find(
            (option) => option.value === event.currentTarget.value
        );
        onChange({ target: { ...option } });
    }
    return (
        <select data-testid="select-portfolio" value={value} onChange={handleChange}>
            {options.map(({ label, value }) => (
                <option key={value} value={value}>
                    {label}
                </option>
            ))}
        </select>
    );
});
