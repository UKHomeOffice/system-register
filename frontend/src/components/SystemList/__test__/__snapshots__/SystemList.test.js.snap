// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SystemList /> matches snapshot 1`] = `
<div
  className="centerContent"
>
  <h1>
    System search
  </h1>
  <div
    className="topMarginBig"
  >
    <strong>
      Number of systems identified: 
    </strong>
    3
  </div>
  <div
    className="bottomMargin topMarginBig"
  >
    <div
      className="MuiFormControl-root MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
        htmlFor="search"
        id="search-label"
      >
        search
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          id="search"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
        />
      </div>
    </div>
    <select
      data-testid="select-portfolio"
      onChange={[Function]}
      value="all"
    >
      <option
        value="Portfolio 1"
      />
      <option
        value="Portfolio 2"
      />
      <option
        value="Portfolio 3"
      />
    </select>
  </div>
  <div
    className="topMarginBig"
  >
    <div
      className="systemCard topMargin"
    >
      <a
        href="/system/undefined"
        onClick={[Function]}
      >
        <strong>
          System 1
        </strong>
        <small>
           (Portfolio: 
          Portfolio 1
          , 
          <span
            data-testid="modified-on"
          >
            Last modified: 
            <span>
              Never
            </span>
          </span>
          )
        </small>
        <div
          style={
            Object {
              "marginTop": "3px",
            }
          }
        >
          <small>
            No Description
          </small>
        </div>
      </a>
    </div>
    <div
      className="systemCard topMargin"
    >
      <a
        href="/system/undefined"
        onClick={[Function]}
      >
        <strong>
          System 3
        </strong>
        <small>
           (Portfolio: 
          Portfolio 2
          , 
          <span
            data-testid="modified-on"
          >
            Last modified: 
            <span>
              Never
            </span>
          </span>
          )
        </small>
        <div
          style={
            Object {
              "marginTop": "3px",
            }
          }
        >
          <small>
            No Description
          </small>
        </div>
      </a>
    </div>
    <div
      className="systemCard topMargin"
    >
      <a
        href="/system/undefined"
        onClick={[Function]}
      >
        <strong>
          System 2
        </strong>
        <small>
           (Portfolio: 
          Portfolio 2
          , 
          <span
            data-testid="modified-on"
          >
            Last modified: 
            <span>
              Never
            </span>
          </span>
          )
        </small>
        <div
          style={
            Object {
              "marginTop": "3px",
            }
          }
        >
          <small>
            No Description
          </small>
        </div>
      </a>
    </div>
  </div>
</div>
`;
